### PR TITLE
Allow FileTaskHandler to delegate to instances of logging.Handler

### PR DESCRIPTION
**Pull request by michalc**
FileTaskHandler is the base class for logging handlers, including those that don't log to files via delegating to logging.FileHandler, e.g. in the CloudwatchTaskHandler at https://github.com/apache/airflow/blob/2940b9fa55a6a72c60c2162e541631addec3d6b8/airflow/providers/amazon/aws/log/cloudwatch_task_handler.py#L67

It is suspected that type checking is not enabled in this part of the CloudwatchTaskHandler otherwise it would have already been failing.

This change adjusts the base class so if type checking is enabled in the task handler, if it delegates to a logging.Handler that is not a logging.FileHandler as the CloudWatchHandler, then the type checking should pass.

This was originally part of https://github.com/apache/airflow/pull/31855 and split out.

related: https://github.com/apache/airflow/issues/31834

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by pierrejeambrun — APPROVED** [review]

File: airflow/providers/microsoft/azure/log/wasb_task_handler.py [hunk 1]
```diff
@@ -62,6 +63,7 @@ def __init__(
         **kwargs,
     ) -> None:
         super().__init__(base_log_folder, filename_template)
+        self.handler: logging.FileHandler | None = None
```
Comment (eladkal) [inline airflow/providers/microsoft/azure/log/wasb_task_handler.py]: Can you undo this change?
it's part of https://github.com/apache/airflow/pull/31855/files

In this PR we want to modify only the core part (not the providers/microsoft)

**Review by eladkal — CHANGES_REQUESTED** [review]

File: airflow/providers/microsoft/azure/log/wasb_task_handler.py (hunk 1 above)
Comment (michalc) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: Ah - without this, type checking fails, so it seemed appropriate otherwise checks would be failing?

But also, https://github.com/apache/airflow/pull/31855/files has just now been merged, so this is a ~mute~ moot point? That PR was rebased on this one

Comment (michalc) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: In hindsight/for future reference, maybe it should have been 3 PRs so checks passed in all of them?

1. One for this line in wasb_task_hander.py
2. One for the change in file_task_handler.py, rebased on 1., then rebased on main after 1. was merged
3. One for the redis task handler, rebased on 2., then rebased on main after 2. was merged?

?

Comment (pierrejeambrun) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: @eladkal is right, usually we try to split core and provider PR because it makes it harder for the release manager.

In cases where it is not possible to do so, then we don't split and try to resolve things at cherry picking time. This can prove difficult and require many PRs from the provider to be manually cherry picked to allow to cherry pick the PR with core change. If it's too much effort, it can postpone the release of this patch until next minor.

Comment (eladkal) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: yeah I think now it should be fine with just the core as the provider code is already updated in main branch
@michalc can you update the PR to drop the `wasb_task_handler.py` change?

Comment (michalc) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: @eladkal Done

Comment (michalc) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: (But have deliberately not rebased on main because I _think_ it will reduce this PR to nothing?)

Comment (eladkal) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: ah oops
I didn't notice `airflow/utils/log/file_task_handler.py` was not removed from https://github.com/apache/airflow/pull/31855 when I merged it.

So NVM... the release manager will handle it as bulk.. :)

**Comment by eladkal** [comment]
Closing as the at the end the change is merged with https://github.com/apache/airflow/pull/31855/

thank you @michalc

File: airflow/providers/microsoft/azure/log/wasb_task_handler.py (hunk 1 above)
Comment (michalc) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: Ah in that case I’ll close this PR?

Comment (michalc) [reply airflow/providers/microsoft/azure/log/wasb_task_handler.py]: Ah I see just done
